File: java/spaceport/src/test/java/com/rocketninesolutions/LaunchedFlightsTest.java
package com.rocketninesolutions;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.RepeatedTest;
import org.junit.jupiter.api.Test;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class LaunchedFlightsTest {

    // These tests attempt to show Launched flights are removed from the departure board
    // Since the SpaceBoardDepartureBoard is using the current time, via LocalDateTime.now()
    // these tests sometimes fail
    // TODO - introduce a mock for the clock to always test against a fixed time

    @Test
    public void launched_WellUnderWindowToRemove_Should_RemainInList() {
        LocalDateTime fourMinutesAgo = LocalDateTime.now().minusMinutes(4);
        LaunchInfo li = createLaunchedInfoWithTime(fourMinutesAgo);

        MockLaunchInfoProvider mock = new MockLaunchInfoProvider();
        mock.launchList.add(li);

        SpaceportDepartureBoard sut = new SpaceportDepartureBoard(mock);
        sut.removeInactiveLaunches();

        Assertions.assertEquals(1, sut.getLaunchList().size(), "Launched four minutes ago, still be on the board");
    }

    @Test
    public void launched_OneSecondUnderWindowToRemove_Should_RemainInList() {
        LocalDateTime almostFiveMinutesAgo = LocalDateTime.now().minusMinutes(5).plusSeconds(1);
        LaunchInfo li = createLaunchedInfoWithTime(almostFiveMinutesAgo);

        MockLaunchInfoProvider mock = new MockLaunchInfoProvider();
        mock.launchList.add(li);

        SpaceportDepartureBoard sut = new SpaceportDepartureBoard(mock);
        sut.removeInactiveLaunches();

        Assertions.assertEquals(1, sut.getLaunchList().size(), "Launch four minutes 59 seconds ago, should still be on the board");
    }

    @RepeatedTest(1000)
    public void launched_LessThanOneSecondUnderWindowToRemove_Should_RemainInList() {
        LocalDateTime almostFiveMinutesAgo = LocalDateTime.now().minusMinutes(5).plusNanos(50000);
        LaunchInfo li = createLaunchedInfoWithTime(almostFiveMinutesAgo);

        MockLaunchInfoProvider mock = new MockLaunchInfoProvider();
        mock.launchList.add(li);

        SpaceportDepartureBoard sut = new SpaceportDepartureBoard(mock);
        sut.removeInactiveLaunches();

        Assertions.assertEquals(1, sut.getLaunchList().size(), "Launched less than 5 minutes ago, should always appear in the list");
    }

    @Test
    public void launched_ExactlyFiveMinutesAgo_ShouldNot_RemainInList() {
        LocalDateTime exactlyFiveMinutesAgo = LocalDateTime.now().minusMinutes(5);
        LaunchInfo li = createLaunchedInfoWithTime(exactlyFiveMinutesAgo);

        MockLaunchInfoProvider mock = new MockLaunchInfoProvider();
        mock.launchList.add(li);

        SpaceportDepartureBoard sut = new SpaceportDepartureBoard(mock);
        sut.removeInactiveLaunches();

        Assertions.assertEquals(0, sut.getLaunchList().size(), "Launched 5 minutes ago, should be removed from list of launches");
    }



    private LaunchInfo createLaunchedInfoWithTime(LocalDateTime launchTime) {
        LaunchInfo li = new LaunchInfo(UUID.randomUUID());
        li.setStatus(LaunchInfo.LaunchStatus.Launched);
        li.setTime(launchTime);
        return li;
    }

    class MockLaunchInfoProvider implements ISpacelineLaunchInfoProvider {

        List<LaunchInfo> launchList;

        public MockLaunchInfoProvider() {
            this.launchList = new ArrayList<>();
        }

        @Override
        public List<LaunchInfo> getCurrentLaunches() {
            return launchList;
        }
    }
}

File: csharp/spaceport/Spaceport/LaunchedFlightsTest.cs
using System;
using System.Collections.Generic;
using System.Threading;
using NUnit.Framework;

namespace SpacePort
{
    [TestFixture]
    public class LaunchedFlightsTest
    {
        // These tests attempt to show Launched flights are removed from the departure board
        // Since the SpaceBoardDepartureBoard is using the current time, via LocalDateTime.now()
        // these tests sometimes fail
        // TODO - introduce a mock for the clock to always test against a fixed time

        [Test]
        public void Launched_WellUnderWindowToRemove_Should_RemainInList()
        {
            DateTime fourMinutesAgo = DateTime.Now.Subtract(new TimeSpan(0, 0, 4, 0));
            LaunchInfo li = CreateLaunchInfoWithTime(fourMinutesAgo);

            MockLaunchInfoProvider mock = new MockLaunchInfoProvider();
            mock.LaunchList.Add(li);

            SpaceportDepartureBoard sut = new SpaceportDepartureBoard(mock);
            sut.RemoveInactiveLaunches();

            Assert.AreEqual(1, sut.LaunchList.Count,"Launched four minutes ago, should still be on the board");
        }

        [Test]
        public void Launched_OneSecondUnderWindowToRemove_Should_RemainInList()
        {
            DateTime almostFiveMinutesAgo = DateTime.Now.Subtract(new TimeSpan(0, 0, 4, 59));
            LaunchInfo li = CreateLaunchInfoWithTime(almostFiveMinutesAgo);

            MockLaunchInfoProvider mock = new MockLaunchInfoProvider();
            mock.LaunchList.Add(li);

            SpaceportDepartureBoard sut = new SpaceportDepartureBoard(mock);
            sut.RemoveInactiveLaunches();

            Assert.AreEqual(1, sut.LaunchList.Count, "Launched four minutes 59 seconds ago, should still be on the board");
        }

        [Test]
        [Repeat(1000)]
        public void Launched_LessThanOneSecondUnderWindowToRemove_Should_RemainInList()
        {
            DateTime almostFiveMinutesAgo = DateTime.Now.Subtract(new TimeSpan(0, 0, 5, 0)).AddMilliseconds(10);
            LaunchInfo li = CreateLaunchInfoWithTime(almostFiveMinutesAgo);
            
            MockLaunchInfoProvider mock = new MockLaunchInfoProvider();
            mock.LaunchList.Add(li);

            SpaceportDepartureBoard sut = new SpaceportDepartureBoard(mock);
            sut.RemoveInactiveLaunches();

            Assert.AreEqual(1, sut.LaunchList.Count, "Launched less than 5 minutes ago, should still be on the board");
        }
        
        [Test]
        public void Launched_ExactlyFiveMinutesAgo_ShouldNot_RemainInList()
        {
            DateTime exactlyFiveMinutesAgo = DateTime.Now.Subtract(new TimeSpan(0, 5, 0));
            LaunchInfo li = CreateLaunchInfoWithTime(exactlyFiveMinutesAgo);

            MockLaunchInfoProvider mock = new MockLaunchInfoProvider();
            mock.LaunchList.Add(li);

            SpaceportDepartureBoard sut = new SpaceportDepartureBoard(mock);
            sut.RemoveInactiveLaunches();

            Assert.AreEqual(0, sut.LaunchList.Count, "Launched 5 minutes ago, should be removed from list of launches");
        }
        
        private class MockLaunchInfoProvider: ISpacelineLaunchInfoProvider
        {
            private List<LaunchInfo> _launchList;

            public List<LaunchInfo> LaunchList
            {
                get => _launchList;
                set => _launchList = value;
            }

            public MockLaunchInfoProvider()
            {
                this._launchList = new List<LaunchInfo>();
                Thread.Sleep(10);   // causes enough delay to interfere with DateTime.Now 
            }
            
            public List<LaunchInfo> GetCurrentLaunches()
            {
                return _launchList;
            }
        }

        private LaunchInfo CreateLaunchInfoWithTime(DateTime time)
        {
            LaunchInfo launchInfo = new LaunchInfo(Guid.NewGuid());
            launchInfo.Status = LaunchInfo.LaunchStatus.Launched;
            launchInfo.Time = time;
            return launchInfo;
        }
    }
}